src/freenet/support/compress/Compressor.java

```diff
@@ -132,22 +132,25 @@ public static void getCompressorDescriptor(StringBuilder sb) {
 				}
 				result.add(ct);
 			}
-			return result.toArray(new COMPRESSOR_TYPE[result.size()]);
+			return result.toArray(new COMPRESSOR_TYPE[0]);
 		}
 
 		@Override
-		public Bucket compress(Bucket data, BucketFactory bf, long maxReadLength, long maxWriteLength) throws IOException, CompressionOutputSizeException {
+		public Bucket compress(Bucket data, BucketFactory bf, long maxReadLength, long maxWriteLength)
+				throws IOException, CompressionOutputSizeException, CompressionRatioException {
 			return compressor.compress(data, bf, maxReadLength, maxWriteLength);
 		}
 
 		@Override
-		public long compress(InputStream is, OutputStream os, long maxReadLength, long maxWriteLength) throws IOException, CompressionOutputSizeException {
+		public long compress(InputStream is, OutputStream os, long maxReadLength, long maxWriteLength)
```

Comment:
This method shouldn't throw `CompressionRatioException`. But the next one should. The next method should be [described](https://github.com/redwerk/fred/commit/26389c1f152b086a6449d7b7f95a432e620e4270#diff-127ee35d8f28c09f0b9ba01bcd477492R202) in a different way. After this not the best [realization](https://github.com/redwerk/fred/commit/26389c1f152b086a6449d7b7f95a432e620e4270#diff-dddbae15c941315e0069a6f233029e27R11) of their interaction might happen.

Another implementation approach: to add some callback and to make a decision about compression continuation at higher level. But in this case a mechanism of compression process will also be needed. It may cause more code without any benefit.

It seems to okey when `compress` with `minimumCompressionPercentage` throw exceptions because it should compress data at least with the set effect, but cannot do this.